### PR TITLE
Sphinx auto docs

### PR DESCRIPTION
**Pull request by jrbenny35**
This is about done. I think it's time we all can add some input on it. 

r? @davehunt

### CODE REVIEW

File: foxpuppet/foxpuppet.py [hunk 1]
```diff
@@ -8,7 +8,11 @@
 
 
 class FoxPuppet(object):
-    """ Class that sets up the api for interacting with the Firefox browser.
+    """
+        Class that sets up the interface for interacting with the Firefox
+        browser.
+
+        :param selenium: A selenium object
```
Comment (davehunt) [inline foxpuppet/foxpuppet.py]: Let's call this a WebDriver object and link to Selenium's docs by including the type. https://seleniumhq.github.io/selenium/docs/api/py/webdriver_remote/selenium.webdriver.remote.webdriver.html#module-selenium.webdriver.remote.webdriver

File: foxpuppet/windows/base.py [hunk 2]
```diff
@@ -8,6 +8,13 @@
 
 class BaseWindow(object):
 
+    """
+        A base window model.
+
+        :param selenium: A selenium object.
```
Comment (davehunt) [inline foxpuppet/windows/base.py]: As above, call this WebDriver, and include the appropriate type.

File: foxpuppet/windows/base.py [hunk 3]
```diff
@@ -8,6 +8,13 @@
...
+        :param handle: A window handle.
```
Comment (davehunt) [inline foxpuppet/windows/base.py]: Add the type (String).

File: foxpuppet/windows/base.py [hunk 4]
```diff
@@ -20,16 +27,17 @@ def document_element(self):
         """ Returns the inner DOM window element.
 
         :returns: DOM window element.
+        :return type: A selenium locator
```
Comment (davehunt) [inline foxpuppet/windows/base.py]: This doesn't return the locator, it returns a WebElement: https://seleniumhq.github.io/selenium/docs/api/py/webdriver_remote/selenium.webdriver.remote.webelement.html#module-selenium.webdriver.remote.webelement

File: foxpuppet/windows/base.py [hunk 5]
```diff
@@ -20,16 +27,17 @@ def document_element(self):
...
 
     def close(self):
-        """Closes the window"""
+        """Closes the window."""
         self.switch_to()
         self.selenium.close()
 
     def switch_to(self):
-        """Switches to the window"""
+        """Switches selenium focus to the window."""
```
Comment (davehunt) [inline foxpuppet/windows/base.py]: Let's reword this to: "Switch focus for Selenium commands to this window."

File: foxpuppet/windows/browser.py [hunk 6]
```diff
@@ -10,7 +10,10 @@
 
 class BrowserWindow(BaseWindow):
 
-    """Representation of a browser window."""
+    """Representation of a browser window.
+
+        :extends: :class:`foxpuppet.windows.BaseWindow`
```
Comment (davehunt) [inline foxpuppet/windows/browser.py]: As `BaseWindow` is not documented, let's exclude this and just allow inherited methods to be included.

File: foxpuppet/windows/browser.py [hunk 7]
```diff
@@ -20,7 +23,9 @@ class BrowserWindow(BaseWindow):
 
     @property
     def is_private(self):
-        """Returns True if this is a Private Browsing window."""
+        """:returns: True if this is a Private Browsing window.
```
Comment (davehunt) [inline foxpuppet/windows/browser.py]: We should include a summary for this property.

File: foxpuppet/windows/browser.py [hunk 8]
```diff
@@ -37,6 +42,9 @@ def open_window(self, private=False):
 
         :param private: Optional parameter to open a private browsing window.
                         Defaults to False.
```
Comment (davehunt) [inline foxpuppet/windows/browser.py]: Include the type.

File: foxpuppet/windows/browser.py [hunk 9]
```diff
@@ -37,6 +42,9 @@ def open_window(self, private=False):
...
+
+        :returns: A BrowserWindow object of the newly opened window.
+        :return type: object
```
Comment (davehunt) [inline foxpuppet/windows/browser.py]: Use the correct type. It should become a link in the docs.

File: foxpuppet/windows/manager.py [hunk 10]
```diff
@@ -5,14 +5,24 @@
 
 class WindowManager(object):
 
+    """
+        A window manager that controls the creation of window objects
+        for interaction.
+
+        :param selenium: A selenium object.
```
Comment (davehunt) [inline foxpuppet/windows/manager.py]: Use the WebDriver type as mentioned earlier.

File: foxpuppet/windows/manager.py [hunk 11]
```diff
@@ -5,14 +5,24 @@
...
         self.selenium = selenium
 
     @property
     def windows(self):
         """
-        Sets all current window handles to appropriate window instances
-        :returns: A list of BrowserWindow Instances
+        Sets all current windows to appropriate window instances.
+
+        :returns: A list of Window Instances.
```
Comment (davehunt) [inline foxpuppet/windows/manager.py]: `BaseWindow` objects

File: foxpuppet/windows/manager.py [hunk 12]
```diff
@@ -5,14 +5,24 @@
...
+
     def __init__(self, selenium):
         self.selenium = selenium
 
     @property
     def windows(self):
         """
-        Sets all current window handles to appropriate window instances
-        :returns: A list of BrowserWindow Instances
+        Sets all current windows to appropriate window instances.
```
Comment (davehunt) [inline foxpuppet/windows/manager.py]: Properties are getters, not setters, so reword this to something like: "Returns a list of all open windows"

File: foxpuppet/foxpuppet.py (hunk 1 above)
Comment (jrbenny35) [inline foxpuppet/foxpuppet.py]: I could use some help with the intersphinx mappings. For some reason it won't work and it appears that I am doing everything right

**Comment by jrbenny35** [comment]
I think I messed up the rebase again. 😭 I will get it one day.

**Comment by jrbenny35** [comment]
@davehunt Could you take a look?

File: foxpuppet/windows/browser/window.py [hunk 13]
```diff
@@ -12,7 +12,8 @@
 
 class BrowserWindow(BaseWindow):
 
-    """Representation of a browser window."""
+    """Representation of a browser window.
```
Comment (davehunt) [inline foxpuppet/windows/browser/window.py]: We need to include the constructor parameters in the docs. I'm guessing that this docstring is taken priority, so we could either look into adding the docstring to the `__init__` method, or just duplicate the parameters in this docstring.

File: foxpuppet/windows/base.py [hunk 14]
```diff
@@ -35,11 +47,11 @@ def firefox_version(self):
...
 
     def close(self):
-        """Closes the window"""
+        """Closes the window."""
         self.switch_to()
         self.selenium.close()
 
     def switch_to(self):
-        """Switches to the window"""
+        """Switches focus for selenium commands to this window."""
```
Comment (davehunt) [inline foxpuppet/windows/base.py]: s/selenium/Selenium/

File: docs/api.rst [hunk 15]
```diff
@@ -0,0 +1,34 @@
...
+
+This part of the documentation describes the interfaces for using FoxPuppet.
+
+FoxPuppet
+---------
+
+.. py:module:: foxpuppet.foxpuppet
+
+.. autoclass:: FoxPuppet
+   :inherited-members:
```
Comment (davehunt) [inline docs/api.rst]: This doesn't subclass anything that we need to inherit.

File: docs/api.rst [hunk 16]
```diff
@@ -0,0 +1,34 @@
...
+
+This module contains all of the window types as well as the window manager.
+
+Window Manager
+==============
+
+.. py:module:: foxpuppet.windows
+
+.. autoclass:: WindowManager
+  :inherited-members:
```
Comment (davehunt) [inline docs/api.rst]: This doesn't subclass anything that we need to inherit.

File: docs/user_guide.rst [hunk 17]
```diff
@@ -1,2 +1,25 @@
...
+    selenium = Firefox()
+    foxpuppet = FoxPuppet(selenium)
+
+    window = foxpuppet.browser.open_window(private=True)
+    selenium.quit()
+
+This example opens first a non-private browser window, then proceeds to open a private browser window.
+Finally it quits via :py:func:`selenium.quit`.
+
+:Note: The initial browser window is automatically assigned to the :py:attr:`browser` attribute and is always available.
```
Comment (davehunt) [inline docs/user_guide.rst]: We already added the user guide - can you remove this change?

Comment (jrbenny35) [inline docs/user_guide.rst]: Whoops!

File: foxpuppet/windows/browser/window.py (hunk 13 above)
Comment (jrbenny35) [inline foxpuppet/windows/browser/window.py]: The constructor parameters of the base class?

Comment (davehunt) [inline foxpuppet/windows/browser/window.py]: Yes, at the moment the generated docs do not detail the parameters for constructing a `BrowserWindow` object.

Comment (jrbenny35) [inline foxpuppet/windows/browser/window.py]: Should we add some kind of note that you can't instantiate this class directly?

Comment (davehunt) [inline foxpuppet/windows/browser/window.py]: Hmm.. that's actually a good point. We probably don't want to document the constructor then! 😄

Comment (jrbenny35) [inline foxpuppet/windows/browser/window.py]: We are taking care of the ```BaseWindow``` creation behind the scenes so it shouldn't affect the end user.

File: docs/user_guide.rst [hunk 18]
```diff
@@ -1,25 +1,2 @@
...
-    selenium = Firefox()
-    foxpuppet = FoxPuppet(selenium)
-
-    window = foxpuppet.browser.open_window(private=True)
-    selenium.quit()
-
-This example opens first a non-private browser window, then proceeds to open a private browser window.
-Finally it quits via :py:func:`selenium.quit`.
-
-:Note: The initial browser window is automatically assigned to the :py:attr:`browser` attribute and is always available.
```
Comment (davehunt) [inline docs/user_guide.rst]: hmm.. we don't want to delete the user guide! 😆

**Comment by jrbenny35** [comment]
@davehunt You told me to remove the user guide since we already have one.
> We already added the user guide - can you remove this change?
